libs/gltfio/src/ReusableAnimator.cpp

```diff
@@ -0,0 +1,521 @@
+            nextIndex = times.size() - 1;
+            prevIndex = nextIndex;
+        } else if (iter == times.begin()) {
+            nextIndex = 0;
+            prevIndex = 0;
+        } else {
+            TimeValues::const_iterator prev = iter; --prev;
+            nextIndex = iter->second;
+            prevIndex = prev->second;
+            const float nextTime = iter->first;
+            const float prevTime = prev->first;
+            float deltaTime = nextTime - prevTime;
+            assert(deltaTime >= 0);
+            if (deltaTime > 0) {
+                t = (time - prevTime) / deltaTime;
+            }
+        }
+
+        if (sampler->interpolation == Sampler::STEP) {
+            t = 0.0f;
+        }
+
+        mImpl->applyAnimation(channel, t, prevIndex, nextIndex);
+    }
+}
+
+void ReusableAnimator::updateBoneMatrices() {
+    auto renderableManager = mImpl->renderableManager;
+    auto transformManager = mImpl->transformManager;
+
+    auto update = [=](const SkinVector& skins, BoneVector& boneVector) {
+        for (const auto& skin : skins) {
+            size_t njoints = skin.joints.size();
+            boneVector.resize(njoints);
+            for (const auto& entity : skin.targets) {
+                auto renderable = renderableManager->getInstance(entity);
+                if (!renderable) {
+                    continue;
+                }
+                mat4f inverseGlobalTransform;
+                auto xformable = transformManager->getInstance(entity);
+                if (xformable) {
+                    inverseGlobalTransform = inverse(transformManager->getWorldTransform(xformable));
+                }
+                for (size_t boneIndex = 0; boneIndex < njoints; ++boneIndex) {
+                    const auto& joint = skin.joints[boneIndex];
+                    TransformManager::Instance jointInstance = transformManager->getInstance(joint);
+                    mat4f globalJointTransform = transformManager->getWorldTransform(jointInstance);
+                    boneVector[boneIndex] =
+                            inverseGlobalTransform *
+                            globalJointTransform *
+                            skin.inverseBindMatrices[boneIndex];
+                }
+                renderableManager->setBones(renderable, boneVector.data(), boneVector.size());
+            }
+        }
+    };
+
+    update(mImpl->asset->mSkins, mImpl->boneMatrices);
+}
+
+float ReusableAnimator::getAnimationDuration(size_t animationIndex) const {
+    return mImpl->animations[animationIndex].duration;
+}
+
+const char* ReusableAnimator::getAnimationName(size_t animationIndex) const {
+    return mImpl->animations[animationIndex].name.c_str();
+}
+
+void ReusableAnimator::clear() {
+    if (mImpl->asset) {
+        for (auto iter : mImpl->zeroTransform) {
+            auto ci = mImpl->transformManager->getInstance(iter.first);
+            mImpl->transformManager->setTransform(ci, iter.second);
```

Comment:
as we remove the animation, we need to return the animated entities to their original transform